airflow/providers/google/cloud/sensors/cloud_composer.py

```diff
@@ -117,3 +128,162 @@ def execute_complete(self, context: dict[str, Any], event: dict[str, str] | None
+        Service Account Token Creator IAM role to the directly preceding identity, with first
+        account from the list granting this role to the originating account (templated).
+    :param poll_interval: Optional: Control the rate of the poll for the result of deferrable run.
+    :param deferrable: Run sensor in deferrable mode
+    """
+
+    template_fields = (
+        "project_id",
+        "region",
+        "environment_id",
+        "composer_dag_id",
+        "impersonation_chain",
+    )
+
+    def __init__(
+        self,
+        *,
+        project_id: str,
+        region: str,
+        environment_id: str,
+        composer_dag_id: str,
+        allowed_states: Iterable[str] | None = None,
+        execution_range: timedelta | list[datetime] | None = None,
+        gcp_conn_id: str = "google_cloud_default",
+        impersonation_chain: str | Sequence[str] | None = None,
+        deferrable: bool = conf.getboolean("operators", "default_deferrable", fallback=False),
+        poll_interval: int = 10,
+        **kwargs,
+    ) -> None:
+        super().__init__(**kwargs)
+        self.project_id = project_id
+        self.region = region
+        self.environment_id = environment_id
+        self.composer_dag_id = composer_dag_id
+        self.allowed_states = list(allowed_states) if allowed_states else [TaskInstanceState.SUCCESS.value]
+        self.execution_range = execution_range
+        self.gcp_conn_id = gcp_conn_id
+        self.impersonation_chain = impersonation_chain
+        self.deferrable = deferrable
+        self.poll_interval = poll_interval
+
+    def _get_execution_dates(self, context):
+        if isinstance(self.execution_range, timedelta):
+            if self.execution_range < timedelta(0):
+                return context["logical_date"], context["logical_date"] - self.execution_range
+            else:
+                return context["logical_date"] - self.execution_range, context["logical_date"]
+        elif isinstance(self.execution_range, list) and len(self.execution_range) > 0:
+            return self.execution_range[0], self.execution_range[1] if len(
+                self.execution_range
+            ) > 1 else context["logical_date"]
+        else:
+            return context["logical_date"] - timedelta(1), context["logical_date"]
+
+    def poke(self, context: Context) -> bool:
+        start_date, end_date = self._get_execution_dates(context)
+
+        if datetime.now(end_date.tzinfo) < end_date:
+            return False
+
+        dag_runs = self._pull_dag_runs()
+
+        self.log.info("Sensor waits for allowed states: %s", self.allowed_states)
+        allowed_states_status = self._check_dag_runs_states(
+            dag_runs=dag_runs,
+            start_date=start_date,
+            end_date=end_date,
+        )
+
+        return allowed_states_status
+
+    def _pull_dag_runs(self) -> list[dict]:
+        """Pull the list of dag runs."""
+        hook = CloudComposerHook(
```

Comment:
IMHO it is better make the hook a property, or assign it to `self` somewhere else, otherwise this hook will be instantiated multiple times, because the sensor will call the `poke()` method multiple times until it reaches the desired state.